### PR TITLE
Fix const in async regression

### LINKED ISSUE #21893: `const` in async no longer works
### Description

```nim
import asyncdispatch

proc test(): Future[bool] {.async.} =
  const S4 = @[byte('a')]
  echo S4 
  return true
```

This was introduced by #20933 which triggers #19849 and prevents https://github.com/status-im/nim-chronos/pull/349

### Nim Version

nim devel, 1.6

### Current Output

```
$ ~/src/Nim/bin/nim c xx
.........................................................................................................................................
/tmp/xx.nim(3, 29) template/generic instantiation of `async` from here
/tmp/xx.nim(4, 9) Error: inconsistent typing for reintroduced symbol 'S4': previous type was: seq[byte]; new type is: array[0..0, byte]
```

### Expected Output

_No response_

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by ire4ever1190**
Closes #21893 

While this doesn't solve the root issue that was brought up, it atleast fixes `const` inside async procs.
It does this by using `typeof` to instead check whether the code has an implicit return or not.

### CODE REVIEW

File: lib/pure/asyncmacro.nim [hunk 1]
```diff
@@ -221,13 +221,11 @@ proc asyncSingleProc(prc: NimNode): NimNode =
...
     procBody.add quote do:
-      template nimAsyncDispatchSetResult(x: `subRetType`) {.used.} =
-        # If the proc has implicit return then this will get called
-        `resultIdent` = x
-      template nimAsyncDispatchSetResult(x: untyped) {.used.} =
-        # If the proc doesn't have implicit return then this will get called
-        x
-    procBody.add newCall(ident"nimAsyncDispatchSetResult", blockStmt)
+      # Check whether there is an implicit return
+      when typeof(`blockStmt`) is void:
```
Comment (arnetheduck) [inline lib/pure/asyncmacro.nim:225]: does it matter that the AST essentially has 3 references to the same AST here? ie the cgen has an ugly habit to modify the AST for example which potentially could be brittle.

**Comment by Araq** [comment]
Which commit caused this regression? @ringabout ?

File: lib/pure/asyncmacro.nim (hunk 1 above)
Comment (ire4ever1190) [reply lib/pure/asyncmacro.nim:225]: Shouldn't the `when` cause only one reference to get to the cgen stage?

**Comment by ringabout** [comment]
> This was introduced by https://github.com/nim-lang/Nim/pull/20933 which triggers https://github.com/nim-lang/Nim/issues/19849

The root cause might be https://github.com/nim-lang/Nim/issues/19849 which is not a regression. But https://github.com/nim-lang/Nim/pull/20933 is the direct cause anyway, which is a regression, that we can revert if time runs out.

File: lib/pure/asyncmacro.nim (hunk 1 above)
Comment (Araq) [reply lib/pure/asyncmacro.nim:225]: The cgen doesn't modify the AST. Well it does and we need to get rid of that hack, but it doesn't modify the AST in a way that hurts here.

**Pull request merged**